### DISCUSSION TITLE
is there a way to get Task status (multi core setup)?

### DISCUSSION THREAD

**Discussion by rubenfrolic** [espressif/arduino-esp32#7478] [Q&A] [answered]
I'm having a setup where I'm running my main code loop an ESP32-S3 on core 1. In parallel I occasionally have a simple LED animation which I'm running on Core 0 using :  xTaskCreatePinnedToCore(). 

After each LED animation I suspend the task, something like this (simplified): 

```
TaskHandle_t LEDS;
xTaskCreatePinnedToCore(updateLed, "LEDS", 10000,NULL,1,&LEDS,0);
```

```
void updateLed(void* pvParameters) {
for (;;)
{
        digitalWrite(LED_RED, HIGH);
        vTaskDelay(1000 / portTICK_PERIOD_MS );
        digitalWrite(LED_RED, LOW);
        vTaskSuspend(NULL);
}
}
```

then when I want to play the LED animation again I simple call 
```
  vTaskResume(LEDS);
```
This works well, and runs smooth. 

However in my main application I want to put the ESP in deep sleep, but only want to do that once my LED animation is finished and the task is suspended. 

Is there an easy way to check the status of the task so I can wait with putting the ESP in deep sleep once the task is done and suspended? 

conceptually something like this:
```
if (vTask(LEDS) == suspended) {
 go to sleep()
}

else{
wait()
}
```

**Comment by mrengineer7777** [comment]
That's interesting.  vTaskSuspend is in a FREETOS header included the ESP32 IDF:
https://github.com/espressif/arduino-esp32/blob/master/tools/sdk/esp32/include/freertos/include/freertos/task.h

I see a function that may satisfy your needs:
```
void vTaskGetInfo( TaskHandle_t xTask,
                   TaskStatus_t * pxTaskStatus,
                   BaseType_t xGetFreeStackSpace,
                   eTaskState eState ) PRIVILEGED_FUNCTION;
```

**Comment by rubenfrolic** [comment]
I'm trying to get it to work, but I'm getting a compilation error. this is what I have: 

```
TaskHandle_t LEDS;
TaskStatus_t xTaskDetails;

void setup() {
  Serial.begin(115200);

  xTaskCreatePinnedToCore(
    updateLed, /* Task function. */
    "LEDS",    /* name of task. */
    10000,     /* Stack size of task */
    NULL,      /* parameter of the task */
    1,         /* priority of the task */
    &LEDS, /* Task handle to keep track of created task */
    0);    /* pin task to core 0 */


   vTaskGetInfo(LEDS,
               &xTaskDetails,
               pdTRUE, // Include the high water mark in xTaskDetails.
               eInvalid ); // Include the task state in xTaskDetails.
}
```


but getting this error:
```
./../../../xtensa-esp32s3-elf/bin/ld: sketch/led_control.ino.cpp.o:(.literal._Z5setupv+0x28): undefined reference to `vTaskGetInfo'
/../../../../xtensa-esp32s3-elf/bin/ld: sketch/led_control.ino.cpp.o: in function `setup()':
/var/folders/v4/1yb3620d48b8689v_ndmlvkc0000gn/T/arduino_modified_sketch_724270/led_control.ino:30: undefined reference to `vTaskGetInfo'
```

I found some examples and I believe I implemented it correctly (but most likely not): 
- [https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/freertos.html](url)
- [https://cpp.hotexamples.com/examples/-/-/vTaskGetInfo/cpp-vtaskgetinfo-function-examples.html](url)

**Comment by bertmelis** [comment]
It's probably because it's not included in the precompiled arduino code.
alternatively, you can use task notifications: https://freertos.org/ulTaskNotifyTake.html

pseudocode because it lacks all the arguments:
```
void updateLed(void* pvParameters) {
for (;;)
{
        digitalWrite(LED_RED, HIGH);
        vTaskDelay(1000 / portTICK_PERIOD_MS );
        digitalWrite(LED_RED, LOW);
        ulTaskNotifyGive()
        vTaskSuspend(NULL);
        ulTaskNotifyTake()
}
}

```
you can check with ulTaskNotifyTake() in your other task. Use `0` for `xTicksToWait` here.

**Comment by rubenfrolic** [comment]
thanks,  will look into this

**Comment by mrengineer7777** [comment]
Found another function `eTaskState eTaskGetState( TaskHandle_t xTask ) PRIVILEGED_FUNCTION;`

**Comment by mrengineer7777** [comment]
Hmmm, vTaskGetInfo() does not want to compile.

Access to the function can be guaranteed with:
```
#include "sdkconfig.h"
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
```
However the function may not exist in the compiled Arduino core.  Using a Semaphore may be the way to go.

**Comment by mrengineer7777** [comment, marked as answer]
`eTaskState ts = eTaskGetState(LEDS);` does compile!

**Reply by rubenfrolic** [reply]
This indeed works!

Some good references for implementation:
- [https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/freertos.html#_CPPv413eTaskGetState12TaskHandle_t](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/freertos.html#_CPPv413eTaskGetState12TaskHandle_t)
- [https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/freertos.html#_CPPv410eTaskState](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/freertos.html#_CPPv410eTaskState)

And this is useful to interpreted what it returns:
- [https://www.freertos.org/RTOS-task-states.html](https://www.freertos.org/RTOS-task-states.html)
